From fe6617cca6b78177e75a056d23cbb5ceb58065c8 Mon Sep 17 00:00:00 2001
From: Dinolek <git@dinolek.me>
Date: Mon, 28 Apr 2025 01:07:25 +0200
Subject: [PATCH 009/115] clk: mediatek: Add main clocks drivers for MT6785

Add drivers for the main system clock controllers present in the
MediaTek MT6785 SoC.
This adds support for the PLLs and top/infra muxes and gates.

Signed-off-by: Dinolek <git@dinolek.me>
---
 drivers/clk/mediatek/Kconfig              |   8 +
 drivers/clk/mediatek/Makefile             |   1 +
 drivers/clk/mediatek/clk-mt6785-apmixed.c | 197 +++++++
 drivers/clk/mediatek/clk-mt6785-infra.c   | 226 ++++++++
 drivers/clk/mediatek/clk-mt6785-top.c     | 638 ++++++++++++++++++++++
 5 files changed, 1070 insertions(+)
 create mode 100644 drivers/clk/mediatek/clk-mt6785-apmixed.c
 create mode 100644 drivers/clk/mediatek/clk-mt6785-infra.c
 create mode 100644 drivers/clk/mediatek/clk-mt6785-top.c

diff --git a/drivers/clk/mediatek/Kconfig b/drivers/clk/mediatek/Kconfig
index fd8440122..10f61eb0a 100644
--- a/drivers/clk/mediatek/Kconfig
+++ b/drivers/clk/mediatek/Kconfig
@@ -303,6 +303,14 @@ config COMMON_CLK_MT6779_AUDSYS
 	help
 	  This driver supports Mediatek MT6779 audsys clocks.
 
+config COMMON_CLK_MT6785
+	tristate "Clock driver for MediaTek MT6785"
+	depends on ARM64 || COMPILE_TEST
+	select COMMON_CLK_MEDIATEK
+	default ARM64
+	help
+	  This driver supports MediaTek MT6785 basic clocks.
+
 config COMMON_CLK_MT6795
 	tristate "Clock driver for MediaTek MT6795"
 	depends on ARCH_MEDIATEK || COMPILE_TEST
diff --git a/drivers/clk/mediatek/Makefile b/drivers/clk/mediatek/Makefile
index d8736a060..b147866a5 100644
--- a/drivers/clk/mediatek/Makefile
+++ b/drivers/clk/mediatek/Makefile
@@ -23,6 +23,7 @@ obj-$(CONFIG_COMMON_CLK_MT6779_VDECSYS) += clk-mt6779-vdec.o
 obj-$(CONFIG_COMMON_CLK_MT6779_VENCSYS) += clk-mt6779-venc.o
 obj-$(CONFIG_COMMON_CLK_MT6779_MFGCFG) += clk-mt6779-mfg.o
 obj-$(CONFIG_COMMON_CLK_MT6779_AUDSYS) += clk-mt6779-aud.o
+obj-$(CONFIG_COMMON_CLK_MT6785) += clk-mt6785-apmixed.o clk-mt6785-infra.o clk-mt6785-top.o
 obj-$(CONFIG_COMMON_CLK_MT6795) += clk-mt6795-apmixedsys.o clk-mt6795-infracfg.o \
 				   clk-mt6795-pericfg.o clk-mt6795-topckgen.o
 obj-$(CONFIG_COMMON_CLK_MT6795_MFGCFG) += clk-mt6795-mfg.o
diff --git a/drivers/clk/mediatek/clk-mt6785-apmixed.c b/drivers/clk/mediatek/clk-mt6785-apmixed.c
new file mode 100644
index 000000000..c107f0c00
--- /dev/null
+++ b/drivers/clk/mediatek/clk-mt6785-apmixed.c
@@ -0,0 +1,197 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2025 Dinolek <git@dinolek.me>
+ */
+
+#include <dt-bindings/clock/mediatek,mt6785-clk.h>
+#include <linux/platform_device.h>
+
+#include "clk-gate.h"
+#include "clk-mtk.h"
+#include "clk-pll.h"
+
+static const struct mtk_gate_regs apmixed_cg_regs = {
+	.sta_ofs = 0x20,
+	.clr_ofs = 0x20,
+	.set_ofs = 0x20,
+};
+
+#define GATE_APMIXED_FLAGS(_id, _name, _parent, _shift, _flags)       \
+	GATE_MTK_FLAGS(_id, _name, _parent, &apmixed_cg_regs, _shift, \
+		       &mtk_clk_gate_ops_no_setclr_inv, _flags)
+
+#define GATE_APMIXED(_id, _name, _parent, _shift) \
+	GATE_APMIXED_FLAGS(_id, _name, _parent, _shift, 0)
+
+static const struct mtk_gate apmixed_clks[] = {
+	GATE_APMIXED(CLK_APMIXED_SSUSB, "apmixed_ssusb", "clk26m", 4),
+	GATE_APMIXED_FLAGS(CLK_APMIXED_APPLL, "apmixed_appll", "clk26m", 5,
+			   CLK_IS_CRITICAL),
+	GATE_APMIXED(CLK_APMIXED_MIPIC0, "apmixed_mipic0", "clk26m", 6),
+	GATE_APMIXED(CLK_APMIXED_MDPLL, "apmixed_mdpll", "clk26m", 7),
+	GATE_APMIXED(CLK_APMIXED_MMSYS, "apmixed_mmsys", "clk26m", 8),
+	GATE_APMIXED(CLK_APMIXED_UFS, "apmixed_ufs", "clk26m", 9),
+	GATE_APMIXED(CLK_APMIXED_MIPIC1, "apmixed_mipic1", "clk26m", 11),
+	GATE_APMIXED(CLK_APMIXED_MEMPLL, "apmixed_mempll", "clk26m", 13),
+	GATE_APMIXED(CLK_APMIXED_LVPLL, "apmixed_lvpll", "clk26m", 14),
+	GATE_APMIXED(CLK_APMIXED_MIPID0, "apmixed_mipid0", "clk26m", 16),
+	GATE_APMIXED(CLK_APMIXED_MIPID1, "apmixed_mipid1", "clk26m", 17),
+};
+
+#define MT6785_PLL_FMAX (3800UL * MHZ)
+#define MT6785_PLL_FMIN (1500UL * MHZ)
+#define MT6785_INTEGER_BITS 8
+
+#define PLL_B(_id, _name, _reg, _pwr_reg, _flags, _rst_bar_mask, _pcwbits,  \
+	      _pd_reg, _pd_shift, _tuner_reg, _tuner_en_reg, _tuner_en_bit, \
+	      _pcw_reg, _div_table)                                         \
+	{                                                                   \
+		.id = _id,                                                  \
+		.name = _name,                                              \
+		.reg = _reg,                                                \
+		.pwr_reg = _pwr_reg,                                        \
+		.pd_reg = _pd_reg,                                          \
+		.tuner_reg = _tuner_reg,                                    \
+		.tuner_en_reg = _tuner_en_reg,                              \
+		.tuner_en_bit = _tuner_en_bit,                              \
+		.pd_shift = _pd_shift,                                      \
+		.flags = _flags,                                            \
+		.rst_bar_mask = _rst_bar_mask,                              \
+		.fmin = MT6785_PLL_FMIN,                                    \
+		.fmax = MT6785_PLL_FMAX,                                    \
+		.pcwbits = _pcwbits,                                        \
+		.pcwibits = MT6785_INTEGER_BITS,                            \
+		.pcw_reg = _pcw_reg,                                        \
+		.div_table = _div_table,                                    \
+	}
+
+#define PLL(_id, _name, _reg, _pwr_reg, _flags, _rst_bar_mask, _pcwbits,    \
+	    _pd_reg, _pd_shift, _tuner_reg, _tuner_en_reg, _tuner_en_bit,   \
+	    _pcw_reg)                                                       \
+	PLL_B(_id, _name, _reg, _pwr_reg, _flags, _rst_bar_mask, _pcwbits,  \
+	      _pd_reg, _pd_shift, _tuner_reg, _tuner_en_reg, _tuner_en_bit, \
+	      _pcw_reg, NULL)
+
+static const struct mtk_pll_div_table armpll_ll_div_table[] = {
+	{ .div = 0, .freq = MT6785_PLL_FMAX },
+	{ .div = 1, .freq = 1700 * MHZ },
+	{ .div = 2, .freq = 850 * MHZ },
+	{ .div = 3, .freq = 425 * MHZ },
+	{ .div = 4, .freq = 212500000 },
+	{ /* sentinel */ }
+};
+
+static const struct mtk_pll_div_table armpll_bl_div_table[] = {
+	{ .div = 0, .freq = MT6785_PLL_FMAX },
+	{ .div = 1, .freq = 1600 * MHZ },
+	{ .div = 2, .freq = 800 * MHZ },
+	{ .div = 3, .freq = 400 * MHZ },
+	{ .div = 4, .freq = 200 * MHZ },
+	{ /* sentinel */ }
+};
+
+static const struct mtk_pll_div_table mfgpll_div_table[] = {
+	{ .div = 0, .freq = MT6785_PLL_FMAX },
+	{ .div = 1, .freq = 1900 * MHZ },
+	{ .div = 2, .freq = 950 * MHZ },
+	{ .div = 3, .freq = 375 * MHZ },
+	{ .div = 4, .freq = 187500000 },
+	{ /* sentinel */ }
+};
+
+static const struct mtk_pll_data plls[] = {
+	PLL_B(CLK_APMIXED_ARMPLL_LL, "armpll_ll", 0x200, 0x20c, PLL_AO, 0, 22,
+	      0x204, 24, 0, 0, 0, 0x204, armpll_ll_div_table),
+	PLL_B(CLK_APMIXED_ARMPLL_BL, "armpll_bl", 0x210, 0x21c, PLL_AO, 0, 22,
+	      0x214, 24, 0, 0, 0, 0x214, armpll_bl_div_table),
+	PLL(CLK_APMIXED_MAINPLL, "mainpll", 0x230, 0x23c, HAVE_RST_BAR | PLL_AO,
+	    BIT(24), 22, 0x234, 24, 0, 0, 0, 0x234),
+	PLL(CLK_APMIXED_UNIV2PLL, "univ2pll", 0x240, 0x24c, HAVE_RST_BAR,
+	    BIT(24), 22, 0x244, 24, 0, 0, 0, 0x244),
+	PLL_B(CLK_APMIXED_MFGPLL, "mfgpll", 0x250, 0x25c, 0, 0, 22, 0x254, 24,
+	      0, 0, 0, 0x254, mfgpll_div_table),
+	PLL(CLK_APMIXED_MSDCPLL, "msdcpll", 0x260, 0x26c, 0, 0, 22, 0x264, 24,
+	    0, 0, 0, 0x264),
+	PLL(CLK_APMIXED_TVDPLL, "tvdpll", 0x270, 0x27c, 0, 0, 22, 0x274, 24, 0,
+	    0, 0, 0x274),
+	PLL(CLK_APMIXED_MMPLL, "mmpll", 0x280, 0x28c, HAVE_RST_BAR, BIT(23), 22,
+	    0x284, 24, 0, 0, 0, 0x284),
+	PLL(CLK_APMIXED_CCIPLL, "ccipll", 0x2a0, 0x2ac, PLL_AO, 0, 22, 0x2a4,
+	    24, 0, 0, 0, 0x2a4),
+	PLL(CLK_APMIXED_ADSPPLL, "adsppll", 0x2b0, 0x2bc, HAVE_RST_BAR, BIT(23),
+	    22, 0x2b4, 24, 0, 0, 0, 0x2b4),
+	PLL(CLK_APMIXED_APLL1, "apll1", 0x2c0, 0x2d0, 0, 0, 32, 0x2c0, 1, 0,
+	    0x14, 0, 0x2c4),
+	PLL(CLK_APMIXED_APLL2, "apll2", 0x2d4, 0x2e4, 0, 0, 32, 0x2d4, 1, 0,
+	    0x14, 1, 0x2d8),
+	PLL(CLK_APMIXED_APUPLL, "apupll", 0x2ec, 0x2f8, 0, 0, 22, 0x2f0, 24, 0,
+	    0, 0, 0x2f0),
+};
+
+static int clk_mt6785_apmixed_probe(struct platform_device *pdev)
+{
+	struct clk_hw_onecell_data *clk_data;
+	struct device_node *node = pdev->dev.of_node;
+	int r;
+
+	clk_data = mtk_alloc_clk_data(CLK_APMIXED_NR_CLK);
+	if (!clk_data)
+		return -ENOMEM;
+
+	r = mtk_clk_register_plls(&pdev->dev, plls, ARRAY_SIZE(plls), clk_data);
+	if (r)
+		goto free_apmixed_data;
+
+	r = mtk_clk_register_gates(&pdev->dev, node, apmixed_clks,
+				   ARRAY_SIZE(apmixed_clks), clk_data);
+	if (r)
+		goto unregister_plls;
+
+	r = of_clk_add_hw_provider(node, of_clk_hw_onecell_get, clk_data);
+	if (r)
+		goto unregister_gates;
+
+	platform_set_drvdata(pdev, clk_data);
+
+	return 0;
+
+unregister_gates:
+	mtk_clk_unregister_gates(apmixed_clks, ARRAY_SIZE(apmixed_clks),
+				 clk_data);
+unregister_plls:
+	mtk_clk_unregister_plls(plls, ARRAY_SIZE(plls), clk_data);
+free_apmixed_data:
+	mtk_free_clk_data(clk_data);
+	return r;
+}
+
+static void clk_mt6785_apmixed_remove(struct platform_device *pdev)
+{
+	struct device_node *node = pdev->dev.of_node;
+	struct clk_hw_onecell_data *clk_data = platform_get_drvdata(pdev);
+
+	of_clk_del_provider(node);
+	mtk_clk_unregister_gates(apmixed_clks, ARRAY_SIZE(apmixed_clks),
+				 clk_data);
+	mtk_clk_unregister_plls(plls, ARRAY_SIZE(plls), clk_data);
+	mtk_free_clk_data(clk_data);
+}
+
+static const struct of_device_id of_match_clk_mt6785_apmixed[] = {
+	{ .compatible = "mediatek,mt6785-apmixedsys" },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, of_match_clk_mt6785_apmixed);
+
+static struct platform_driver clk_mt6785_apmixed_drv = {
+	.probe = clk_mt6785_apmixed_probe,
+	.remove = clk_mt6785_apmixed_remove,
+	.driver = {
+		.name = "clk-mt6785-apmixed",
+		.of_match_table = of_match_clk_mt6785_apmixed,
+	},
+};
+module_platform_driver(clk_mt6785_apmixed_drv);
+
+MODULE_DESCRIPTION("MediaTek MT6785 apmixed clocks driver");
+MODULE_LICENSE("GPL");
diff --git a/drivers/clk/mediatek/clk-mt6785-infra.c b/drivers/clk/mediatek/clk-mt6785-infra.c
new file mode 100644
index 000000000..3688b7af1
--- /dev/null
+++ b/drivers/clk/mediatek/clk-mt6785-infra.c
@@ -0,0 +1,226 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2025 Dinolek <git@dinolek.me>
+ */
+
+#include <dt-bindings/clock/mediatek,mt6785-clk.h>
+#include <dt-bindings/reset/mediatek,mt6785-resets.h>
+#include <linux/platform_device.h>
+
+#include "clk-gate.h"
+#include "clk-mtk.h"
+
+static const struct mtk_gate_regs infra0_cg_regs = {
+	.sta_ofs = 0x90,
+	.clr_ofs = 0x84,
+	.set_ofs = 0x80,
+};
+
+static const struct mtk_gate_regs infra1_cg_regs = {
+	.sta_ofs = 0x94,
+	.clr_ofs = 0x8c,
+	.set_ofs = 0x88,
+};
+
+static const struct mtk_gate_regs infra2_cg_regs = {
+	.sta_ofs = 0xac,
+	.clr_ofs = 0xa8,
+	.set_ofs = 0xa4,
+};
+
+static const struct mtk_gate_regs infra3_cg_regs = {
+	.sta_ofs = 0xc8,
+	.clr_ofs = 0xc4,
+	.set_ofs = 0xc0,
+};
+
+#define GATE_INFRA0(_id, _name, _parent, _shift)               \
+	GATE_MTK(_id, _name, _parent, &infra0_cg_regs, _shift, \
+		 &mtk_clk_gate_ops_setclr)
+
+#define GATE_INFRA1(_id, _name, _parent, _shift)               \
+	GATE_MTK(_id, _name, _parent, &infra1_cg_regs, _shift, \
+		 &mtk_clk_gate_ops_setclr)
+
+#define GATE_INFRA2(_id, _name, _parent, _shift)               \
+	GATE_MTK(_id, _name, _parent, &infra2_cg_regs, _shift, \
+		 &mtk_clk_gate_ops_setclr)
+
+#define GATE_INFRA2_FLAGS(_id, _name, _parent, _shift, _flag)        \
+	GATE_MTK_FLAGS(_id, _name, _parent, &infra2_cg_regs, _shift, \
+		       &mtk_clk_gate_ops_setclr, _flag)
+
+#define GATE_INFRA3(_id, _name, _parent, _shift)               \
+	GATE_MTK(_id, _name, _parent, &infra3_cg_regs, _shift, \
+		 &mtk_clk_gate_ops_setclr)
+
+#define GATE_INFRA3_FLAGS(_id, _name, _parent, _shift, _flag)        \
+	GATE_MTK_FLAGS(_id, _name, _parent, &infra3_cg_regs, _shift, \
+		       &mtk_clk_gate_ops_setclr, _flag)
+
+static const struct mtk_gate infra_clks[] = {
+	GATE_INFRA0(CLK_INFRA_PMIC_TMR, "infra_pmic_tmr", "top_pwrap_ulposc", 0),
+	GATE_INFRA0(CLK_INFRA_PMIC_AP, "infra_pmic_ap", "top_pwrap_ulposc", 1),
+	GATE_INFRA0(CLK_INFRA_PMIC_MD, "infra_pmic_md", "top_pwrap_ulposc", 2),
+	GATE_INFRA0(CLK_INFRA_PMIC_CONN, "infra_pmic_conn", "top_pwrap_ulposc", 3),
+	GATE_INFRA0(CLK_INFRA_SCP, "infra_scp", "top_scp", 4),
+	GATE_INFRA0(CLK_INFRA_SEJ, "infra_sej", "clk26m", 5),
+	GATE_INFRA0(CLK_INFRA_APXGPT, "infra_apxgpt", "top_axi", 6),
+	GATE_INFRA0(CLK_INFRA_ICUSB, "infra_icusb", "top_axi", 8),
+	GATE_INFRA0(CLK_INFRA_GCE, "infra_gce", "top_axi", 9),
+	GATE_INFRA0(CLK_INFRA_THERM, "infra_therm", "top_axi", 10),
+	GATE_INFRA0(CLK_INFRA_I2C_AP, "infra_i2c_ap", "top_i2c", 11),
+	GATE_INFRA0(CLK_INFRA_I2C_CCU, "infra_i2c_ccu", "top_i2c", 12),
+	GATE_INFRA0(CLK_INFRA_I2C_SSPM, "infra_i2c_sspm", "top_i2c", 13),
+	GATE_INFRA0(CLK_INFRA_I2C_RSV, "infra_i2c_rsv", "top_i2c", 14),
+	GATE_INFRA0(CLK_INFRA_PWM_HCLK, "infra_pwm_hclk", "top_pwm", 15),
+	GATE_INFRA0(CLK_INFRA_PWM1, "infra_pwm1", "top_pwm", 16),
+	GATE_INFRA0(CLK_INFRA_PWM2, "infra_pwm2", "top_pwm", 17),
+	GATE_INFRA0(CLK_INFRA_PWM3, "infra_pwm3", "top_pwm", 18),
+	GATE_INFRA0(CLK_INFRA_PWM4, "infra_pwm4", "top_pwm", 19),
+	GATE_INFRA0(CLK_INFRA_PWM, "infra_pwm", "top_pwm", 21),
+	GATE_INFRA0(CLK_INFRA_UART0, "infra_uart0", "top_uart", 22),
+	GATE_INFRA0(CLK_INFRA_UART1, "infra_uart1", "top_uart", 23),
+	GATE_INFRA0(CLK_INFRA_UART2, "infra_uart2", "top_uart", 24),
+	GATE_INFRA0(CLK_INFRA_UART3, "infra_uart3", "top_uart", 25),
+	GATE_INFRA0(CLK_INFRA_GCE_26M, "infra_gce_26m", "top_axi", 27),
+	GATE_INFRA0(CLK_INFRA_CQ_DMA_FPC, "infra_cq_dma_fpc", "top_axi", 28),
+	GATE_INFRA0(CLK_INFRA_BTIF, "infra_btif", "top_axi", 31),
+
+	GATE_INFRA1(CLK_INFRA_SPI0, "infra_spi0", "top_spi", 1),
+	GATE_INFRA1(CLK_INFRA_MSDC0, "infra_msdc0", "top_msdc50_0_hclk", 2),
+	GATE_INFRA1(CLK_INFRA_MSDC1, "infra_msdc1", "top_axi", 4),
+	GATE_INFRA1(CLK_INFRA_MSDC0_SRC, "infra_msdc0_src", "top_msdc50_0", 6),
+	GATE_INFRA1(CLK_INFRA_DVFSRC, "infra_dvfsrc", "clk26m", 7),
+	GATE_INFRA1(CLK_INFRA_GCPU, "infra_gcpu", "top_axi", 8),
+	GATE_INFRA1(CLK_INFRA_TRNG, "infra_trng", "top_axi", 9),
+	GATE_INFRA1(CLK_INFRA_AUXADC, "infra_auxadc", "clk26m", 10),
+	GATE_INFRA1(CLK_INFRA_CPUM, "infra_cpum", "top_axi", 11),
+	GATE_INFRA1(CLK_INFRA_CCIF1_AP, "infra_ccif1_ap", "top_axi", 12),
+	GATE_INFRA1(CLK_INFRA_CCIF1_MD, "infra_ccif1_md", "top_axi", 13),
+	GATE_INFRA1(CLK_INFRA_AUXADC_MD, "infra_auxadc_md", "clk26m", 14),
+	GATE_INFRA1(CLK_INFRA_MSDC1_SRC, "infra_msdc1_src", "top_msdc30_1", 16),
+	GATE_INFRA1(CLK_INFRA_AP_DMA, "infra_ap_dma", "top_axi", 18),
+	GATE_INFRA1(CLK_INFRA_XIU, "infra_xiu", "top_axi", 19),
+	GATE_INFRA1(CLK_INFRA_DEVICE_APC, "infra_device_apc", "top_axi", 20),
+	GATE_INFRA1(CLK_INFRA_CCIF_AP, "infra_ccif_ap", "top_axi", 23),
+	GATE_INFRA1(CLK_INFRA_DEBUGSYS, "infra_debugsys", "top_axi", 24),
+	GATE_INFRA1(CLK_INFRA_AUDIO, "infra_audio", "top_axi", 25),
+	GATE_INFRA1(CLK_INFRA_CCIF_MD, "infra_ccif_md", "top_axi", 26),
+	GATE_INFRA1(CLK_INFRA_DXCC_SEC_CORE, "infra_dxcc_sec_core", "top_dxcc", 27),
+	GATE_INFRA1(CLK_INFRA_DXCC, "infra_dxcc", "top_dxcc", 28),
+	GATE_INFRA1(CLK_INFRA_DEVMPU, "infra_devmpu", "top_axi", 30),
+	GATE_INFRA1(CLK_INFRA_DRAMC_26M, "infra_dramc_26m", "clk26m", 31),
+
+	GATE_INFRA2(CLK_INFRA_IRTX, "infra_irtx", "clk26m", 0),
+	GATE_INFRA2(CLK_INFRA_SSUSB, "infra_ssusb", "top_ssusb", 1),
+	GATE_INFRA2(CLK_INFRA_DISP_PWM, "infra_disppwm", "top_axi", 2),
+	GATE_INFRA2(CLK_INFRA_AUDIO_26M, "infra_audio_26m", "clk26m", 4),
+	GATE_INFRA2(CLK_INFRA_SPI1, "infra_spi1", "top_spi", 6),
+	GATE_INFRA2(CLK_INFRA_I2C4, "infra_i2c4", "top_i2c", 7),
+	GATE_INFRA2(CLK_INFRA_MODEM_TEMP_SHARE, "infra_md_tmp_share", "clk26m", 8),
+	GATE_INFRA2(CLK_INFRA_SPI2, "infra_spi2", "top_spi", 9),
+	GATE_INFRA2(CLK_INFRA_SPI3, "infra_spi3", "top_spi", 10),
+	GATE_INFRA2(CLK_INFRA_UNIPRO_SYS, "infra_unipro_sys", "top_ufs", 11),
+	GATE_INFRA2(CLK_INFRA_UNIPRO_TICK, "infra_unipro_tick", "top_ufs", 12),
+	GATE_INFRA2(CLK_INFRA_UFS_MP_SAP, "infra_ufs_mp_sap", "top_ufs", 13),
+	GATE_INFRA2(CLK_INFRA_MD32, "infra_md32", "top_axi", 14),
+	GATE_INFRA2_FLAGS(CLK_INFRA_SSPM, "infra_sspm", "top_sspm", 15, CLK_IS_CRITICAL),
+	GATE_INFRA2(CLK_INFRA_UNIPRO_MBIST, "infra_unipro_mbist", "top_axi", 16),
+	GATE_INFRA2_FLAGS(CLK_INFRA_SSPM_BUS_HCLK, "infra_sspm_bus_hclk",
+			  "top_axi", 17, CLK_IS_CRITICAL),
+	GATE_INFRA2(CLK_INFRA_I2C5, "infra_i2c5", "top_i2c", 18),
+	GATE_INFRA2(CLK_INFRA_I2C5_ARBITER, "infra_i2c5_arbiter", "top_i2c", 19),
+	GATE_INFRA2(CLK_INFRA_I2C5_IMM, "infra_i2c5_imm", "top_i2c", 20),
+	GATE_INFRA2(CLK_INFRA_I2C1_ARBITER, "infra_i2c1_arbiter", "top_i2c", 21),
+	GATE_INFRA2(CLK_INFRA_I2C1_IMM, "infra_i2c1_imm", "top_i2c", 22),
+	GATE_INFRA2(CLK_INFRA_I2C2_ARBITER, "infra_i2c2_arbiter", "top_i2c", 23),
+	GATE_INFRA2(CLK_INFRA_I2C2_IMM, "infra_i2c2_imm", "top_i2c", 24),
+	GATE_INFRA2(CLK_INFRA_SPI4, "infra_spi4", "top_spi", 25),
+	GATE_INFRA2(CLK_INFRA_SPI5, "infra_spi5", "top_spi", 26),
+	GATE_INFRA2(CLK_INFRA_CQ_DMA, "infra_cq_dma", "top_axi", 27),
+	GATE_INFRA2(CLK_INFRA_UFS, "infra_ufs", "top_ufs", 28),
+	GATE_INFRA2(CLK_INFRA_AES_UFSFDE, "infra_aes_ufsfde", "top_aes_ufsfde", 29),
+	GATE_INFRA2(CLK_INFRA_UFS_TICK, "infra_ufs_tick", "top_ufs", 30),
+	GATE_INFRA2(CLK_INFRA_SSUSB_XHCI, "infra_ssusb_xhci", "top_ssusb_xhci", 31),
+
+	GATE_INFRA3(CLK_INFRA_MSDC0_SELF, "infra_msdc0_self", "top_msdc50_0", 0),
+	GATE_INFRA3(CLK_INFRA_MSDC1_SELF, "infra_msdc1_self", "top_msdc50_0", 1),
+	GATE_INFRA3_FLAGS(CLK_INFRA_SSPM_26M_SELF, "infra_sspm_26m_self",
+			  "clk26m", 3, CLK_IS_CRITICAL),
+	GATE_INFRA3_FLAGS(CLK_INFRA_SSPM_32K_SELF, "infra_sspm_32k_self",
+			  "clk32k", 4, CLK_IS_CRITICAL),
+	GATE_INFRA3(CLK_INFRA_UFS_AXI, "infra_ufs_axi", "top_axi", 5),
+	GATE_INFRA3(CLK_INFRA_I2C6, "infra_i2c6", "top_i2c", 6),
+	GATE_INFRA3(CLK_INFRA_AP_MSDC0, "infra_ap_msdc0", "top_msdc50_0_hclk", 7),
+	GATE_INFRA3(CLK_INFRA_MD_MSDC0, "infra_md_msdc0", "top_msdc50_0_hclk", 8),
+	GATE_INFRA3(CLK_INFRA_ADSP_26M, "infra_adsp_26m", "top_axi", 12),
+	GATE_INFRA3(CLK_INFRA_ADSP_32K, "infra_adsp_32k", "top_axi", 13),
+	GATE_INFRA3(CLK_INFRA_CCIF2_AP, "infra_ccif2_ap", "top_axi", 16),
+	GATE_INFRA3(CLK_INFRA_CCIF2_MD, "infra_ccif2_md", "top_axi", 17),
+	GATE_INFRA3(CLK_INFRA_CCIF3_AP, "infra_ccif3_ap", "top_axi", 18),
+	GATE_INFRA3(CLK_INFRA_CCIF3_MD, "infra_ccif3_md", "top_axi", 19),
+	GATE_INFRA3(CLK_INFRA_SEJ_13M, "infra_sej_13m", "clk13m", 20),
+	GATE_INFRA3(CLK_INFRA_AES, "infra_aes", "top_axi", 21),
+	GATE_INFRA3(CLK_INFRA_I2C7, "infra_i2c7", "top_i2c", 22),
+	GATE_INFRA3(CLK_INFRA_I2C8, "infra_i2c8", "top_i2c", 23),
+	GATE_INFRA3(CLK_INFRA_BIST2FPC, "infra_bist2fpc", "top_msdc50_0", 24),
+	GATE_INFRA3_FLAGS(CLK_INFRA_DEVICE_APC_SYNC, "infra_device_apc_sync",
+			  "top_axi", 25, CLK_IS_CRITICAL),
+	GATE_INFRA3(CLK_INFRA_DPMAIF, "infra_dpmaif", "top_dpmaif", 26),
+	GATE_INFRA3(CLK_INFRA_ADSP, "infra_adsp", "top_adsp", 27),
+	GATE_INFRA3(CLK_INFRA_CCIF4_AP, "infra_ccif4_ap", "top_axi", 28),
+	GATE_INFRA3(CLK_INFRA_CCIF4_MD, "infra_ccif4_md", "top_axi", 29),
+	GATE_INFRA3(CLK_INFRA_SPI6, "infra_spi6", "top_spi", 30),
+	GATE_INFRA3(CLK_INFRA_SPI7, "infra_spi7", "top_spi", 31),
+};
+
+static u16 infra_rst_ofs[] = {
+	INFRA_RST0_SET_OFFSET,
+	INFRA_RST1_SET_OFFSET,
+	INFRA_RST2_SET_OFFSET,
+	INFRA_RST3_SET_OFFSET,
+	INFRA_RST4_SET_OFFSET,
+};
+
+static u16 infra_idx_map[] = {
+	[MT6785_INFRA_THERM_CTRL_RST] = 0 * RST_NR_PER_BANK + 0,
+	[MT6785_INFRA_UFS_HCI_RST] = 1 * RST_NR_PER_BANK + 15,
+	[MT6785_INFRA_PMIC_WRAP_RST] = 2 * RST_NR_PER_BANK + 0,
+	[MT6785_INFRA_UFS_UNIPRO_RST] = 2 * RST_NR_PER_BANK + 7,
+	[MT6785_INFRA_SVS_RST] = 3 * RST_NR_PER_BANK + 5,
+	[MT6785_INFRA_UFS_AES_RST] = 3 * RST_NR_PER_BANK + 21,
+};
+
+static const struct mtk_clk_rst_desc clk_rst_desc = {
+	.version = MTK_RST_SET_CLR,
+	.rst_bank_ofs = infra_rst_ofs,
+	.rst_bank_nr = ARRAY_SIZE(infra_rst_ofs),
+	.rst_idx_map = infra_idx_map,
+	.rst_idx_map_nr = ARRAY_SIZE(infra_idx_map),
+};
+
+static const struct mtk_clk_desc infra_desc = {
+	.clks = infra_clks,
+	.num_clks = ARRAY_SIZE(infra_clks),
+	.rst_desc = &clk_rst_desc,
+};
+
+static const struct of_device_id of_match_clk_mt6785_infra[] = {
+	{ .compatible = "mediatek,mt6785-infracfg", .data = &infra_desc },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, of_match_clk_mt6785_infra);
+
+static struct platform_driver clk_mt6785_infra_drv = {
+	.probe = mtk_clk_simple_probe,
+	.remove = mtk_clk_simple_remove,
+	.driver = {
+		.name = "clk-mt6785-infra",
+		.of_match_table = of_match_clk_mt6785_infra,
+	},
+};
+module_platform_driver(clk_mt6785_infra_drv);
+
+MODULE_DESCRIPTION("MediaTek MT6785 Infrastructure clocks driver");
+MODULE_LICENSE("GPL");
diff --git a/drivers/clk/mediatek/clk-mt6785-top.c b/drivers/clk/mediatek/clk-mt6785-top.c
new file mode 100644
index 000000000..062d8efc7
--- /dev/null
+++ b/drivers/clk/mediatek/clk-mt6785-top.c
@@ -0,0 +1,638 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2025 Dinolek <git@dinolek.me>
+ */
+
+#include <dt-bindings/clock/mediatek,mt6785-clk.h>
+#include <linux/platform_device.h>
+
+#include "clk-gate.h"
+#include "clk-mtk.h"
+#include "clk-mux.h"
+
+#define CLK_TOP_CFG_UPDATE0 0x4
+#define CLK_TOP_CFG_ENTRIES 32
+
+#define CLK_TOP_CFG_0 0x20
+
+#define TOP_MUX_GATE_FLAGS(_mux, _id, _name, _parents, _shift, _width, _gate, \
+			   _upd, _flags)                                      \
+	MUX_GATE_CLR_SET_UPD_FLAGS(                                           \
+		_id, _name, _parents,                                         \
+		(CLK_TOP_CFG_0 + 0x10 * (_mux)),                              \
+		(CLK_TOP_CFG_0 + 0x10 * (_mux) + 0x4),                        \
+		(CLK_TOP_CFG_0 + 0x10 * (_mux) + 0x8),                        \
+		_shift, _width, _gate,                                        \
+		(CLK_TOP_CFG_UPDATE0 + 0x4 * ((_upd) / CLK_TOP_CFG_ENTRIES)), \
+		((_upd) % CLK_TOP_CFG_ENTRIES), _flags                        \
+	)
+
+#define TOP_MUX_GATE(_mux, _id, _name, _parents, _shift, _width, _gate, _upd) \
+	TOP_MUX_GATE_FLAGS(_mux, _id, _name, _parents, _shift, _width, _gate, \
+			   _upd, CLK_SET_RATE_PARENT)
+
+#define TOP_MUX_GATE_CRITICAL(_mux, _id, _name, _parents, _shift, _width,     \
+			      _gate, _upd)                                    \
+	TOP_MUX_GATE_FLAGS(_mux, _id, _name, _parents, _shift, _width, _gate, \
+			   _upd, CLK_IS_CRITICAL | CLK_SET_RATE_PARENT)
+
+static DEFINE_SPINLOCK(mt6785_clk_lock);
+
+static const struct mtk_fixed_clk top_fixed_clks[] = {
+	FIXED_CLK(CLK_TOP_ULPOSC, "ulposc", NULL, 260000000),
+};
+
+static const struct mtk_fixed_factor top_factor_clks[] = {
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D2, "mainpll_d2", "mainpll", 1, 2, 0),
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D2_D2, "mainpll_d2_d2", "mainpll_d2", 1, 2, 0),
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D2_D4, "mainpll_d2_d4", "mainpll_d2", 1, 4, 0),
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D2_D8, "mainpll_d2_d8", "mainpll_d2", 1, 8, 0),
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D2_D16, "mainpll_d2_d16", "mainpll_d2", 1, 16, 0),
+
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D3, "mainpll_d3", "mainpll", 1, 3, 0),
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D3_D2, "mainpll_d3_d2", "mainpll_d3", 1, 2, 0),
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D3_D4, "mainpll_d3_d4", "mainpll_d3", 1, 4, 0),
+
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D5, "mainpll_d5", "mainpll", 1, 5, 0),
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D5_D2, "mainpll_d5_d2", "mainpll_d5", 1, 2, 0),
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D5_D4, "mainpll_d5_d4", "mainpll_d5", 1, 4, 0),
+
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D7, "mainpll_d7", "mainpll", 1, 7, 0),
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D7_D2, "mainpll_d7_d2", "mainpll_d7", 1, 2, 0),
+	FACTOR_FLAGS(CLK_TOP_MAINPLL_D7_D4, "mainpll_d7_d4", "mainpll_d7", 1, 4, 0),
+
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL, "univpll", "univ2pll", 1, 2, 0),
+
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D2, "univpll_d2", "univpll", 1, 2, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D2_D2, "univpll_d2_d2", "univpll_d2", 1, 2, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D2_D4, "univpll_d2_d4", "univpll_d2", 1, 4, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D2_D8, "univpll_d2_d8", "univpll_d2", 1, 8, 0),
+
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D3, "univpll_d3", "univpll", 1, 3, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D3_D2, "univpll_d3_d2", "univpll_d3", 1, 2, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D3_D4, "univpll_d3_d4", "univpll_d3", 1, 4, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D3_D8, "univpll_d3_d8", "univpll_d3", 1, 8, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D3_D16, "univpll_d3_d16", "univpll_d3", 1, 16, 0),
+
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D5, "univpll_d5", "univpll", 1, 5, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D5_D2, "univpll_d5_d2", "univpll_d5", 1, 2, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D5_D4, "univpll_d5_d4", "univpll_d5", 1, 4, 0),
+
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_D7, "univpll_d7", "univpll", 1, 7, 0),
+
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_192M, "univpll_192m", "univ2pll", 1, 13, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_192M_D2, "univpll_192m_d2", "univpll_192m", 1, 2, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_192M_D4, "univpll_192m_d4", "univpll_192m", 1, 4, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_192M_D8, "univpll_192m_d8", "univpll_192m", 1, 8, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_192M_D16, "univpll_192m_d16", "univpll_192m", 1, 16, 0),
+	FACTOR_FLAGS(CLK_TOP_UNIVPLL_192M_D32, "univpll_192m_d32", "univpll_192m", 1, 32, 0),
+
+	FACTOR(CLK_TOP_APLL1_D2, "apll1_d2", "apll1", 1, 2),
+	FACTOR(CLK_TOP_APLL1_D4, "apll1_d4", "apll1", 1, 4),
+	FACTOR(CLK_TOP_APLL1_D8, "apll1_d8", "apll1", 1, 8),
+
+	FACTOR(CLK_TOP_APLL2_D2, "apll2_d2", "apll2", 1, 2),
+	FACTOR(CLK_TOP_APLL2_D4, "apll2_d4", "apll2", 1, 4),
+	FACTOR(CLK_TOP_APLL2_D8, "apll2_d8", "apll2", 1, 8),
+
+	FACTOR(CLK_TOP_TVDPLL_D2, "tvdpll_d2", "tvdpll", 1, 2),
+	FACTOR(CLK_TOP_TVDPLL_D4, "tvdpll_d4", "tvdpll", 1, 4),
+	FACTOR(CLK_TOP_TVDPLL_D8, "tvdpll_d8", "tvdpll", 1, 8),
+	FACTOR(CLK_TOP_TVDPLL_D16, "tvdpll_d16", "tvdpll", 1, 16),
+
+	FACTOR(CLK_TOP_MMPLL_D4, "mmpll_d4", "mmpll", 1, 4),
+	FACTOR(CLK_TOP_MMPLL_D4_D2, "mmpll_d4_d2", "mmpll_d4", 1, 2),
+	FACTOR(CLK_TOP_MMPLL_D5, "mmpll_d5", "mmpll", 1, 5),
+	FACTOR(CLK_TOP_MMPLL_D5_D2, "mmpll_d5_d2", "mmpll_d5", 1, 2),
+	FACTOR(CLK_TOP_MMPLL_D6, "mmpll_d6", "mmpll", 1, 6),
+	FACTOR(CLK_TOP_MMPLL_D7, "mmpll_d7", "mmpll", 1, 7),
+
+	FACTOR(CLK_TOP_ADSPPLL_D4, "adsppll_d4", "adsppll", 1, 4),
+	FACTOR(CLK_TOP_ADSPPLL_D5, "adsppll_d5", "adsppll", 1, 5),
+	FACTOR(CLK_TOP_ADSPPLL_D6, "adsppll_d6", "adsppll", 1, 6),
+
+	FACTOR(CLK_TOP_MSDCPLL_D2, "msdcpll_d2", "msdcpll", 1, 2),
+	FACTOR(CLK_TOP_MSDCPLL_D4, "msdcpll_d4", "msdcpll", 1, 4),
+
+	FACTOR(CLK_TOP_OSC_D2, "osc_d2", "ulposc", 1, 2),
+	FACTOR(CLK_TOP_OSC_D4, "osc_d4", "ulposc", 1, 4),
+	FACTOR(CLK_TOP_OSC_D8, "osc_d8", "ulposc", 1, 8),
+	FACTOR(CLK_TOP_OSC_D10, "osc_d10", "ulposc", 1, 10),
+	FACTOR(CLK_TOP_OSC_D16, "osc_d16", "ulposc", 1, 16),
+
+	FACTOR(CLK_TOP_TVDPLL_MAINPLL_D2, "tvdpll_mainpll_d2", "tvdpll", 1, 1),
+};
+
+static const char *const axi_parents[] = {
+	"clk26m",
+	"mainpll_d2_d4",
+	"mainpll_d7",
+	"osc_d4"
+};
+
+static const char *const mm_parents[] = {
+	"clk26m",
+	"adsppll_d5",
+	"mmpll_d7",
+	"mmpll_d5_d2",
+	"mainpll_d2_d2",
+	"tvdpll_mainpll_d2"
+};
+
+static const char *const scp_parents[] = {
+	"clk26m",
+	"mainpll_d2_d2",
+	"mainpll_d2_d4",
+	"mainpll_d3",
+	"univpll_d3",
+	"mainpll_d5",
+	"mainpll_d3_d2",
+};
+
+static const char *const isp_parents[] = {
+	"clk26m",
+	"mainpll_d2",
+	"mmpll_d6",
+	"adsppll_d5",
+	"mainpll_d3",
+	"mmpll_d5_d2",
+	"tvdpll_mainpll_d2",
+	"mainpll_d2_d2"
+};
+
+static const char *const cam_parents[] = {
+	"clk26m",
+	"mainpll_d2",
+	"mmpll_d6",
+	"mainpll_d3",
+	"mmpll_d7",
+	"univpll_d3",
+	"mmpll_d5_d2",
+	"adsppll_d5",
+	"tvdpll_mainpll_d2",
+	"univpll_d3_d2"
+};
+
+static const char *const ccu_parents[] = {
+	"clk26m",
+	"mainpll_d2",
+	"mmpll_d6",
+	"mainpll_d3",
+	"mmpll_d7",
+	"univpll_d3",
+	"mmpll_d5_d2",
+	"mainpll_d2_d2",
+	"adsppll_d5",
+	"univpll_d3_d2"
+};
+
+static const char *const dsp_ipu_if_parents[] = {
+	"clk26m",
+	"univpll_d3_d8",
+	"univpll_d3_d4",
+	"mainpll_d2_d4",
+	"univpll_d3_d2",
+	"mainpll_d2_d2",
+	"univpll_d2_d2",
+	"mainpll_d3",
+	"univpll_d3",
+	"mmpll_d7",
+	"mmpll_d6",
+	"adsppll_d5",
+	"tvdpll",
+	"univpll_d2",
+	"adsppll_d4",
+	"apupll"
+};
+
+static const char *const mfg_parents[] = {
+	"clk26m",
+	"mfgpll",
+	"univpll_d3",
+	"mainpll_d5"
+};
+
+static const char *const mfg_52m_parents[] = {
+	"clk26m",
+	"univpll_d3_d2",
+	"univpll_d3_d4",
+	"univpll_d3_d8"
+};
+
+static const char *const camtg_parents[] = {
+	"clk26m",
+	"univpll_192m_d8",
+	"univpll_d3_d8",
+	"univpll_192m_d4",
+	"univpll_d3_d16",
+	"clk13m",
+	"univpll_192m_d16",
+	"univpll_192m_d32"
+};
+
+static const char *const uart_parents[] = {
+	"clk26m",
+	"univpll_d3_d8"
+};
+
+static const char *const spi_parents[] = {
+	"clk26m",
+	"mainpll_d5_d2",
+	"mainpll_d3_d4",
+	"msdcpll_d4"
+};
+
+static const char *const msdc50_0_hclk_parents[] = {
+	"clk26m",
+	"mainpll_d2_d2",
+	"mainpll_d3_d2"
+};
+
+static const char *const msdc50_0_parents[] = {
+	"clk26m",
+	"msdcpll",
+	"msdcpll_d2",
+	"univpll_d2_d4",
+	"mainpll_d3_d2",
+	"univpll_d2_d2"
+};
+
+static const char *const msdc30_1_parents[] = {
+	"clk26m",
+	"univpll_d3_d2",
+	"mainpll_d3_d2",
+	"mainpll_d7",
+	"msdcpll_d2"
+};
+
+static const char *const audio_parents[] = {
+	"clk26m",
+	"mainpll_d5_d4",
+	"mainpll_d7_d4",
+	"mainpll_d2_d16"
+};
+
+static const char *const audio_intbus_parents[] = {
+	"clk26m",
+	"mainpll_d2_d4",
+	"mainpll_d7_d2"
+};
+
+static const char *const pwrap_ulposc_parents[] = {
+	"osc_d10",
+	"clk26m",
+	"osc_d4",
+	"osc_d8",
+	"osc_d16"
+};
+
+static const char *const atb_parents[] = {
+	"clk26m",
+	"mainpll_d2_d2",
+	"mainpll_d5"
+};
+
+static const char *const sspm_parents[] = {
+	"clk26m",
+	"univpll_d2_d4",
+	"mainpll_d2_d2",
+	"univpll_d2_d2",
+	"mainpll_d3"
+};
+
+static const char *const dpi_parents[] = {
+	"clk26m",
+	"tvdpll_d2",
+	"tvdpll_d4",
+	"tvdpll_d8",
+	"tvdpll_d16"
+};
+
+static const char *const scam_parents[] = {
+	"clk26m",
+	"mainpll_d5_d2"
+};
+
+static const char *const disp_pwm_parents[] = {
+	"clk26m",
+	"univpll_d3_d4",
+	"osc_d2",
+	"osc_d4",
+	"osc_d16"
+};
+
+static const char *const ssusb_parents[] = {
+	"clk26m",
+	"univpll_d5_d4",
+	"univpll_d3_d4",
+	"univpll_d5_d2"
+};
+
+static const char *const spm_parents[] = {
+	"clk26m",
+	"osc_d8",
+	"mainpll_d2_d8"
+};
+
+static const char *const i2c_parents[] = {
+	"clk26m",
+	"mainpll_d2_d8",
+	"univpll_d5_d2"
+};
+
+static const char *const seninf_parents[] = {
+	"clk26m",
+	"univpll_d7",
+	"univpll_d3_d2",
+	"univpll_d2_d2"
+};
+
+static const char *const dxcc_parents[] = {
+	"clk26m",
+	"mainpll_d2_d2",
+	"mainpll_d2_d4",
+	"mainpll_d2_d8"
+};
+
+static const char *const audio_eng1_parents[] = {
+	"clk26m",
+	"apll1_d2",
+	"apll1_d4",
+	"apll1_d8"
+};
+
+static const char *const audio_eng2_parents[] = {
+	"clk26m",
+	"apll2_d2",
+	"apll2_d4",
+	"apll2_d8"
+};
+
+static const char *const aes_ufsfde_parents[] = {
+	"clk26m",
+	"mainpll_d2",
+	"mainpll_d2_d2",
+	"mainpll_d3",
+	"mainpll_d2_d4",
+	"univpll_d3"
+};
+
+static const char *const ufs_parents[] = {
+	"clk26m",
+	"mainpll_d2_d4",
+	"mainpll_d2_d8",
+	"mainpll_d2_d16"
+};
+
+static const char *const audio1_parents[] = {
+	"clk26m",
+	"apll1"
+};
+
+static const char *const audio2_parents[] = {
+	"clk26m",
+	"apll2"
+};
+
+static const char *const adsp_parents[] = {
+	"clk26m",
+	"mainpll_d3",
+	"univpll_d2_d4",
+	"univpll_d2",
+	"mmpll_d4",
+	"adsppll_d4",
+	"adsppll_d6"
+};
+
+static const char *const dpmaif_parents[] = {
+	"clk26m",
+	"univpll_d2_d4",
+	"mainpll_d3",
+	"mainpll_d2_d2",
+	"univpll_d2_d2",
+	"univpll_d3"
+};
+
+static const char *const venc_parents[] = {
+	"clk26m",
+	"mmpll_d7",
+	"mainpll_d3",
+	"univpll_d2_d2",
+	"mainpll_d2_d2",
+	"univpll_d3",
+	"mmpll_d6",
+	"mainpll_d5",
+	"mainpll_d3_d2",
+	"mmpll_d4_d2",
+	"univpll_d2_d4",
+	"mmpll_d5",
+	"univpll_192m_d2"
+};
+
+static const char *const vdec_parents[] = {
+	"clk26m",
+	"univpll_d2_d4",
+	"mainpll_d3",
+	"univpll_d2_d2",
+	"mainpll_d2_d2",
+	"univpll_d3",
+	"univpll_d5",
+	"univpll_d5_d2",
+	"mainpll_d2",
+	"univpll_d2",
+	"univpll_192m_d2"
+};
+
+static const char *const camtm_parents[] = {
+	"clk26m",
+	"univpll_d7",
+	"univpll_d3_d2",
+	"univpll_d2_d2"
+};
+
+static const char *const pwm_parents[] = {
+	"clk26m",
+	"univpll_d2_d8"
+};
+
+static const char *const audio_hires_parents[] = {
+	"clk26m",
+	"univpll_d7",
+	"apll1",
+	"apll2"
+};
+
+static const char *const apll_i2s_parents[] = {
+	"top_audio1",
+	"top_audio2"
+};
+
+static const struct mtk_mux top_mux_clks[] = {
+	TOP_MUX_GATE_CRITICAL(0, CLK_TOP_AXI, "top_axi", axi_parents, 0, 2, 7, 0),
+	TOP_MUX_GATE(0, CLK_TOP_MM, "top_mm", mm_parents, 8, 3, 15, 1),
+	TOP_MUX_GATE(0, CLK_TOP_SCP, "top_scp", scp_parents, 16, 3, 23, 2),
+	/* cksys_fmem */
+
+	TOP_MUX_GATE(1, CLK_TOP_IMG, "top_img", isp_parents, 0, 3, 7, 4),
+	TOP_MUX_GATE(1, CLK_TOP_IPE, "top_ipe", isp_parents, 8, 3, 15, 5),
+	TOP_MUX_GATE(1, CLK_TOP_DPE, "top_dpe", isp_parents, 16, 3, 23, 6),
+	TOP_MUX_GATE(1, CLK_TOP_CAM, "top_cam", cam_parents, 24, 4, 31, 7),
+
+	TOP_MUX_GATE(2, CLK_TOP_CCU, "top_ccu", ccu_parents, 0, 4, 7, 8),
+	TOP_MUX_GATE(2, CLK_TOP_DSP0, "top_dsp0", dsp_ipu_if_parents, 8, 4, 15, 9),
+	TOP_MUX_GATE(2, CLK_TOP_DSP1, "top_dsp1", dsp_ipu_if_parents, 16, 4, 23, 10),
+	TOP_MUX_GATE(2, CLK_TOP_DSP2, "top_dsp2", dsp_ipu_if_parents, 24, 4, 31, 11),
+
+	/* dsp3 */
+	TOP_MUX_GATE(3, CLK_TOP_IPU_IF, "top_ipu_if", dsp_ipu_if_parents, 8, 4, 15, 13),
+	TOP_MUX_GATE(3, CLK_TOP_MFG, "top_mfg", mfg_parents, 16, 2, 23, 14),
+	TOP_MUX_GATE(3, CLK_TOP_MFG_52M, "top_mfg_52m", mfg_52m_parents, 24, 2, 31, 15),
+
+	TOP_MUX_GATE(4, CLK_TOP_CAMTG1, "top_camtg1", camtg_parents, 0, 3, 7, 16),
+	TOP_MUX_GATE(4, CLK_TOP_CAMTG2, "top_camtg2", camtg_parents, 8, 3, 15, 17),
+	TOP_MUX_GATE(4, CLK_TOP_CAMTG3, "top_camtg3", camtg_parents, 16, 3, 23, 18),
+	TOP_MUX_GATE(4, CLK_TOP_CAMTG4, "top_camtg4", camtg_parents, 24, 3, 31, 19),
+
+	TOP_MUX_GATE(5, CLK_TOP_UART, "top_uart", uart_parents, 0, 1, 7, 20),
+	TOP_MUX_GATE(5, CLK_TOP_SPI, "top_spi", spi_parents, 8, 2, 15, 21),
+	TOP_MUX_GATE_FLAGS(5, CLK_TOP_MSDC50_0_HCLK, "top_msdc50_0_hclk",
+			   msdc50_0_hclk_parents, 16, 2, 23, 22, 0),
+	TOP_MUX_GATE_FLAGS(5, CLK_TOP_MSDC50_0, "top_msdc50_0",
+			   msdc50_0_parents, 24, 3, 31, 23, 0),
+
+	TOP_MUX_GATE_FLAGS(6, CLK_TOP_MSDC30_1, "top_msdc30_1",
+			   msdc30_1_parents, 0, 3, 7, 24, 0),
+	TOP_MUX_GATE(6, CLK_TOP_AUDIO, "top_audio", audio_parents, 8, 2, 15, 25),
+	TOP_MUX_GATE(6, CLK_TOP_AUDIO_INTBUS, "top_audio_intbus",
+		     audio_intbus_parents, 16, 2, 23, 26),
+	TOP_MUX_GATE(6, CLK_TOP_PWRAP_ULPOSC, "top_pwrap_ulposc",
+		     pwrap_ulposc_parents, 24, 3, 31, 27),
+
+	TOP_MUX_GATE(7, CLK_TOP_ATB, "top_atb", atb_parents, 0, 2, 7, 28),
+	TOP_MUX_GATE_CRITICAL(7, CLK_TOP_SSPM, "top_sspm", sspm_parents, 8, 3, 15, 29),
+	TOP_MUX_GATE(7, CLK_TOP_DPI, "top_dpi", dpi_parents, 16, 3, 23, 30),
+	TOP_MUX_GATE(7, CLK_TOP_SCAM, "top_scam", scam_parents, 24, 1, 31, 32),
+
+	TOP_MUX_GATE(8, CLK_TOP_DISP_PWM, "top_disp_pwm", disp_pwm_parents, 0, 3, 7, 33),
+	TOP_MUX_GATE(8, CLK_TOP_SSUSB, "top_ssusb", ssusb_parents, 8, 2, 15, 34),
+	TOP_MUX_GATE(8, CLK_TOP_SSUSB_XHCI, "top_ssusb_xhci", ssusb_parents, 16, 2, 23, 35),
+	TOP_MUX_GATE_CRITICAL(8, CLK_TOP_SPM, "top_spm", spm_parents, 24, 2, 31, 36),
+
+	TOP_MUX_GATE(9, CLK_TOP_I2C, "top_i2c", i2c_parents, 0, 2, 7, 37),
+	TOP_MUX_GATE(9, CLK_TOP_SENINF0, "top_seninf0", seninf_parents, 8, 2, 15, 38),
+	TOP_MUX_GATE(9, CLK_TOP_SENINF1, "top_seninf1", seninf_parents, 16, 2, 23, 39),
+	TOP_MUX_GATE(9, CLK_TOP_SENINF2, "top_seninf2", seninf_parents, 24, 2, 31, 40),
+
+	TOP_MUX_GATE(10, CLK_TOP_DXCC, "top_dxcc", dxcc_parents, 0, 2, 7, 41),
+	TOP_MUX_GATE(10, CLK_TOP_AUDIO_ENG1, "top_audio_eng1",
+		     audio_eng1_parents, 8, 2, 15, 42),
+	TOP_MUX_GATE(10, CLK_TOP_AUDIO_ENG2, "top_audio_eng2",
+		     audio_eng2_parents, 16, 2, 23, 43),
+	TOP_MUX_GATE(10, CLK_TOP_AES_UFSFDE, "top_aes_ufsfde",
+		     aes_ufsfde_parents, 24, 3, 31, 44),
+
+	TOP_MUX_GATE(11, CLK_TOP_UFS, "top_ufs", ufs_parents, 0, 2, 7, 45),
+	TOP_MUX_GATE(11, CLK_TOP_AUDIO1, "top_audio1", audio1_parents, 8, 1, 15, 46),
+	TOP_MUX_GATE(11, CLK_TOP_AUDIO2, "top_audio2", audio2_parents, 16, 1, 23, 47),
+	TOP_MUX_GATE(11, CLK_TOP_ADSP, "top_adsp", adsp_parents, 24, 3, 31, 48),
+
+	TOP_MUX_GATE(12, CLK_TOP_DPMAIF, "top_dpmaif", dpmaif_parents, 0, 3, 7, 49),
+	TOP_MUX_GATE(12, CLK_TOP_VENC, "top_venc", venc_parents, 8, 4, 15, 50),
+	TOP_MUX_GATE(12, CLK_TOP_VDEC, "top_vdec", vdec_parents, 16, 4, 23, 51),
+	TOP_MUX_GATE(12, CLK_TOP_CAMTM, "top_camtm", camtm_parents, 24, 2, 31, 52),
+
+	TOP_MUX_GATE(13, CLK_TOP_PWM, "top_pwm", pwm_parents, 0, 1, 7, 53),
+	TOP_MUX_GATE(13, CLK_TOP_AUDIO_HIRES, "top_audio_hires",
+		     audio_hires_parents, 8, 2, 15, 54),
+	/* bus_aximem */
+	TOP_MUX_GATE(13, CLK_TOP_CAMTG5, "top_camtg5", camtg_parents, 24, 3, 31, 56),
+};
+
+static const struct mtk_composite top_composite_clks[] = {
+	MUX(CLK_TOP_APLL_I2S0, "top_apll_i2s0", apll_i2s_parents, 0x320, 8, 1),
+	MUX(CLK_TOP_APLL_I2S1, "top_apll_i2s1", apll_i2s_parents, 0x320, 9, 1),
+	MUX(CLK_TOP_APLL_I2S2, "top_apll_i2s2", apll_i2s_parents, 0x320, 10, 1),
+	MUX(CLK_TOP_APLL_I2S3, "top_apll_i2s3", apll_i2s_parents, 0x320, 11, 1),
+	MUX(CLK_TOP_APLL_I2S4, "top_apll_i2s4", apll_i2s_parents, 0x320, 12, 1),
+	MUX(CLK_TOP_APLL_I2S5, "top_apll_i2s5", apll_i2s_parents, 0x328, 20, 1),
+
+	DIV_GATE(CLK_TOP_APLL12_DIV0, "top_apll12_div0", "top_apll_i2s0", 0x320, 2, 0x324, 8, 0),
+	DIV_GATE(CLK_TOP_APLL12_DIV1, "top_apll12_div1", "top_apll_i2s1", 0x320, 3, 0x324, 8, 8),
+	DIV_GATE(CLK_TOP_APLL12_DIV2, "top_apll12_div2", "top_apll_i2s2", 0x320, 4, 0x324, 8, 16),
+	DIV_GATE(CLK_TOP_APLL12_DIV3, "top_apll12_div3", "top_apll_i2s3", 0x320, 5, 0x324, 8, 24),
+	DIV_GATE(CLK_TOP_APLL12_DIV4, "top_apll12_div4", "top_apll_i2s4", 0x320, 6, 0x328, 8, 0),
+	DIV_GATE(CLK_TOP_APLL12_DIVB, "top_apll12_divb", "top_apll12_div4", 0x320, 7, 0x328, 8, 8),
+	DIV_GATE(CLK_TOP_APLL12_DIV5, "top_apll12_div5", "top_apll_i2s5", 0x328, 16, 0x328, 4, 28),
+};
+
+static const struct mtk_gate_regs top_cg_regs = {
+	.sta_ofs = 0x110,
+	.clr_ofs = 0x110,
+	.set_ofs = 0x110,
+};
+
+#define GATE_TOP(_id, _name, _parent, _shift)               \
+	GATE_MTK(_id, _name, _parent, &top_cg_regs, _shift, \
+		 &mtk_clk_gate_ops_no_setclr_inv)
+
+static const struct mtk_gate top_clks[] = {
+	GATE_TOP(CLK_TOP_ARMPLL_DIV_PLL1, "armpll_div_pll1", "mainpll", 4),
+	GATE_TOP(CLK_TOP_ARMPLL_DIV_PLL2, "armpll_div_pll2", "univpll", 5),
+};
+
+/* Register mux notifier for MFG mux */
+static int clk_mt6785_reg_mfg_mux_notifier(struct device *dev, struct clk *clk)
+{
+	struct mtk_mux_nb *mfg_mux_nb;
+	int i;
+
+	mfg_mux_nb = devm_kzalloc(dev, sizeof(*mfg_mux_nb), GFP_KERNEL);
+	if (!mfg_mux_nb)
+		return -ENOMEM;
+
+	for (i = 0; i < ARRAY_SIZE(top_mux_clks); i++)
+		if (top_mux_clks[i].id == CLK_TOP_MFG)
+			break;
+	if (i == ARRAY_SIZE(top_mux_clks))
+		return -EINVAL;
+
+	mfg_mux_nb->ops = top_mux_clks[i].ops;
+	mfg_mux_nb->bypass_index = 3; /* Bypass to mainpll_d5 */
+
+	return devm_mtk_clk_mux_notifier_register(dev, clk, mfg_mux_nb);
+}
+
+static const struct mtk_clk_desc top_desc = {
+	.clks = top_clks,
+	.num_clks = ARRAY_SIZE(top_clks),
+	.fixed_clks = top_fixed_clks,
+	.num_fixed_clks = ARRAY_SIZE(top_fixed_clks),
+	.factor_clks = top_factor_clks,
+	.num_factor_clks = ARRAY_SIZE(top_factor_clks),
+	.mux_clks = top_mux_clks,
+	.num_mux_clks = ARRAY_SIZE(top_mux_clks),
+	.composite_clks = top_composite_clks,
+	.num_composite_clks = ARRAY_SIZE(top_composite_clks),
+	.clk_lock = &mt6785_clk_lock,
+	.clk_notifier_func = clk_mt6785_reg_mfg_mux_notifier,
+	.mfg_clk_idx = CLK_TOP_MFG,
+};
+
+static const struct of_device_id of_match_clk_mt6785_top[] = {
+	{ .compatible = "mediatek,mt6785-topckgen", .data = &top_desc },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, of_match_clk_mt6785_top);
+
+static struct platform_driver clk_mt6785_top_drv = {
+	.probe = mtk_clk_simple_probe,
+	.remove = mtk_clk_simple_remove,
+	.driver = {
+		.name = "clk-mt6785-top",
+		.of_match_table = of_match_clk_mt6785_top,
+	},
+};
+module_platform_driver(clk_mt6785_top_drv);
+
+MODULE_DESCRIPTION("MediaTek MT6785 top clock generators driver");
+MODULE_LICENSE("GPL");
-- 
2.56.0

